File: pm_core/src/infrastructure/store/adapter/issue_type_scheme_item_adapter.go
package adapter

import (
	"context"
	"fmt"

	"github.com/serp/pm-core/src/core/domain/entity"
	"github.com/serp/pm-core/src/core/port/store"
	"github.com/serp/pm-core/src/infrastructure/store/mapper"
	"github.com/serp/pm-core/src/infrastructure/store/model"
	"gorm.io/gorm"
)

type IssueTypeSchemeItemAdapter struct {
	db     *gorm.DB
	mapper *mapper.IssueTypeSchemeItemMapper
}

func NewIssueTypeSchemeItemAdapter(db *gorm.DB) store.IIssueTypeSchemeItemStore {
	return &IssueTypeSchemeItemAdapter{
		db:     db,
		mapper: mapper.NewIssueTypeSchemeItemMapper(),
	}
}

func (a *IssueTypeSchemeItemAdapter) CreateIssueTypeSchemeItem(ctx context.Context, tx *gorm.DB, item *entity.IssueTypeSchemeItemEntity) (*entity.IssueTypeSchemeItemEntity, error) {
	db := a.getDB(tx)
	itemModel := a.mapper.ToModel(item)
	if err := db.WithContext(ctx).Create(itemModel).Error; err != nil {
		return nil, fmt.Errorf("failed to create issue type scheme item: %w", err)
	}
	return a.mapper.ToEntity(itemModel), nil
}

func (a *IssueTypeSchemeItemAdapter) CreateIssueTypeSchemeItems(ctx context.Context, tx *gorm.DB, items []*entity.IssueTypeSchemeItemEntity) ([]*entity.IssueTypeSchemeItemEntity, error) {
	if len(items) == 0 {
		return []*entity.IssueTypeSchemeItemEntity{}, nil
	}
	db := a.getDB(tx)
	itemModels := a.mapper.ToModels(items)
	if err := db.WithContext(ctx).CreateInBatches(itemModels, 100).Error; err != nil {
		return nil, fmt.Errorf("failed to create issue type scheme items: %w", err)
	}
	return a.mapper.ToEntities(itemModels), nil
}

func (a *IssueTypeSchemeItemAdapter) DeleteIssueTypeSchemeItemsBySchemeID(ctx context.Context, tx *gorm.DB, schemeID, tenantID int64) error {
	db := a.getDB(tx)
	if err := db.WithContext(ctx).
		Model(&model.IssueTypeSchemeItemModel{}).
		Where("scheme_id = ? AND tenant_id = ? AND deleted_at IS NULL", schemeID, tenantID).
		Update("deleted_at", gorm.Expr("CURRENT_TIMESTAMP")).Error; err != nil {
		return fmt.Errorf("failed to delete issue type scheme items by scheme id: %w", err)
	}
	return nil
}

func (a *IssueTypeSchemeItemAdapter) GetIssueTypeSchemeItemsBySchemeID(ctx context.Context, schemeID, tenantID int64) ([]*entity.IssueTypeSchemeItemEntity, error) {
	var itemModels []*model.IssueTypeSchemeItemModel
	if err := a.db.WithContext(ctx).
		Where("scheme_id = ? AND tenant_id = ? AND deleted_at IS NULL", schemeID, tenantID).
		Order("sequence ASC").
		Find(&itemModels).Error; err != nil {
		return nil, fmt.Errorf("failed to get issue type scheme items by scheme id: %w", err)
	}
	return a.mapper.ToEntities(itemModels), nil
}

func (a *IssueTypeSchemeItemAdapter) ExistsIssueTypeInScheme(ctx context.Context, schemeID, issueTypeID, tenantID int64) (bool, error) {
	var count int64
	if err := a.db.WithContext(ctx).Model(&model.IssueTypeSchemeItemModel{}).
		Where("scheme_id = ? AND issue_type_id = ? AND tenant_id = ? AND deleted_at IS NULL", schemeID, issueTypeID, tenantID).
		Count(&count).Error; err != nil {
		return false, fmt.Errorf("failed to check issue type in scheme: %w", err)
	}
	return count > 0, nil
}

func (a *IssueTypeSchemeItemAdapter) getDB(tx *gorm.DB) *gorm.DB {
	if tx != nil {
		return tx
	}
	return a.db
}

File: pm_core/src/main/java/serp/project/pmcore/infrastructure/store/adapter/IssueTypeSchemeItemAdapter.java
/**
 * Author: QuanTuanHuy
 * Description: Part of Serp Project
 */

package serp.project.pmcore.infrastructure.store.adapter;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import serp.project.pmcore.core.domain.entity.IssueTypeSchemeItemEntity;
import serp.project.pmcore.core.port.store.IIssueTypeSchemeItemPort;
import serp.project.pmcore.infrastructure.store.mapper.IssueTypeSchemeItemMapper;
import serp.project.pmcore.infrastructure.store.repository.IIssueTypeSchemeItemRepository;

import java.util.ArrayList;
import java.util.List;

@Component
@RequiredArgsConstructor
public class IssueTypeSchemeItemAdapter implements IIssueTypeSchemeItemPort {

    private final IIssueTypeSchemeItemRepository issueTypeSchemeItemRepository;
    private final IssueTypeSchemeItemMapper issueTypeSchemeItemMapper;

    @Override
    public IssueTypeSchemeItemEntity createIssueTypeSchemeItem(IssueTypeSchemeItemEntity item) {
        return issueTypeSchemeItemMapper.toEntity(issueTypeSchemeItemRepository.save(issueTypeSchemeItemMapper.toModel(item)));
    }

    @Override
    public List<IssueTypeSchemeItemEntity> createIssueTypeSchemeItems(List<IssueTypeSchemeItemEntity> items) {
        if (items == null || items.isEmpty()) {
            return new ArrayList<>();
        }
        return issueTypeSchemeItemMapper.toEntities(
                issueTypeSchemeItemRepository.saveAll(issueTypeSchemeItemMapper.toModels(items))
        );
    }

    @Override
    public List<IssueTypeSchemeItemEntity> getIssueTypeSchemeItemsBySchemeId(Long schemeId, Long tenantId) {
        return issueTypeSchemeItemMapper.toEntities(
                issueTypeSchemeItemRepository.findAllByTenantIdAndSchemeIdOrderBySequenceAsc(tenantId, schemeId)
        );
    }

    @Override
    public void deleteIssueTypeSchemeItemsBySchemeId(Long schemeId, Long tenantId) {
        issueTypeSchemeItemRepository.deleteBySchemeIdAndTenantId(schemeId, tenantId);
    }

    @Override
    public boolean existsIssueTypeInScheme(Long schemeId, Long issueTypeId, Long tenantId) {
        return issueTypeSchemeItemRepository.existsByTenantIdAndSchemeIdAndIssueTypeId(tenantId, schemeId, issueTypeId);
    }
}
